### PR TITLE
8343705: IGV: Interactive Node Moving in Hierarchical Layout

### PR DESCRIPTION
**Pull request by tobiasholenstein**
This PR depends on https://github.com/openjdk/jdk/pull/22402 . To check out this PR locally:
`git fetch https://git.openjdk.org/jdk.git pull/22430/head:pull/22430`
`git checkout pull/22430`

This pull request enhances the Ideal Graph Visualizer (IGV) by introducing an interactive feature that allows users to move nodes within the hierarchical layout by dragging them to new positions. This manual adjustment helps users better understand and explore the graph structure by customizing the layout according to their needs.

## Overview

Previously, the hierarchical layout in IGV was static, and users could not adjust node positions manually. This limitation made it challenging to reorganize the graph for improved readability or to focus on specific areas of interest. With this enhancement is new:
- Interactive Node Movement: Users can now click and drag nodes to new positions within the graph.
- Dynamic Edge Adjustment: When nodes are moved, connected edges adjust dynamically to maintain the graph's structure.
- Layer Management: Nodes can be moved within the same layer or across different layers, with the layout updating accordingly.
- Persistent Positions: Moved nodes remain in their new positions until the layout is reset or the nodes are moved again.

## Limitations
- Interactive Node Moving only works in `Sea of nodes` view with `cut long edges` off. (The standard option for IGV)
- Currently, no empty layers are allowed in the layout, so users cannot introduce a horizontal gap between nodes that only contains edges.
- To move long straight edges, it's best to drag the top part of the edges around.
When the graph changes - for example, when nodes are removed or hidden, or layers are applied - the rearrangements are lost since the graph gets re-laid out. To preserve rearrangements, support for a stable incremental layout algorithm would be needed.

### LayoutMover Interface
Created a new interface `LayoutMover` with methods `moveVertex`, `moveVertices`, and `moveLink`.
`HierarchicalLayoutManager` now implements `LayoutMover`, providing concrete implementations for these methods.

###  Enhancements to HierarchicalLayoutManager
Improved the `HierarchicalLayoutManager` so it can handle moving nodes interactively. Added methods to move single nodes or multiple nodes, and to adjust links. Nodes can now be moved within the same layer or to different layers while keeping the graph consistent. Also added a `writeBack` method to apply these changes.

### Modifications to LayoutGraph
Updated `LayoutGraph` to support changes when nodes are moved. Added methods to manage layers, like creating and deleting layers, inserting new ones when needed, and removing empty layers. Also added ways to handle nodes and edges, like removing them or adding edges with dummy nodes if necessary. Improved how edges are handled to reduce crossings and straighten them, and made better algorithms to position nodes and minimize edge crossings.

### User Interface Enhancements in DiagramScene
Enhanced the user interface to let users move nodes and edges by dragging them. Added visual feedback like shadows and pointers when moving nodes. We updated the actions for `FigureWidget` to include move actions, making nodes draggable. We also adjusted the methods that rebuild the layout to handle these dynamic changes.

### FigureWidget and LineWidget
Made changes so that when nodes and edges are moved, the UI components update accordingly. In `FigureWidget`, added an `updatePosition` method to update its position when the node moves. In `LineWidget`, we made the start and end points editable so edges can adjust when nodes move, and added methods to support moving edges interactively.

### 3. Moving the position where an edge crosses a layer

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343705](https://bugs.openjdk.org/browse/JDK-8343705): IGV: Interactive Node Moving in Hierarchical Layout (**Enhancement** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) Review applies to [b8444a3a](https://git.openjdk.org/jdk/pull/22430/files/b8444a3adcccb21c6c58b995d096a7b81f1e8a49)
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**) Review applies to [2410ac79](https://git.openjdk.org/jdk/pull/22430/files/2410ac7956a6dfd7ffb352d9d15021020dbd0f8c)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22430/head:pull/22430` \
`$ git checkout pull/22430`

Update a local copy of the PR: \
`$ git checkout pull/22430` \
`$ git pull https://git.openjdk.org/jdk.git pull/22430/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22430`

View PR using the GUI difftool: \
`$ git pr show -t 22430`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22430.diff">https://git.openjdk.org/jdk/pull/22430.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22430#issuecomment-2506035879)
</details>

### CODE REVIEW

**Review by TobiHartmann — APPROVED** [review]
I thoroughly tested this with a few complex graphs and it works great. This is an awesome enhancement!

**Review by robcasloz — COMMENTED** [review]
I tested and benchmarked this change using the same methodology as in [JDK-8314512](https://github.com/openjdk/jdk/pull/22402). Both testing and performance results are OK. The change seems to introduce a slight overhead of around 10% w.r.t. JDK-8314512 but still speeds up the baseline (current IGV) by around 1.4x thanks to disabling assertions in JDK-8314512. See full results here: [results.ods](https://github.com/user-attachments/files/17948626/results.ods). I will start to review the actual code changes.

**Review by robcasloz — APPROVED** [review]
Looks good!

**Review by robcasloz — APPROVED** [review]
I re-tested this change after merging the latest changes from #22402 and did not find any issue.

File: src/utils/IdealGraphVisualizer/HierarchicalLayout/src/main/java/com/sun/hotspot/igv/hierarchicallayout/LayoutGraph.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2008, 2024, Oracle and/or its affiliates. All rights reserved.
```
Comment (chhagedorn) [inline src/utils/IdealGraphVisualizer/HierarchicalLayout/src/main/java/com/sun/hotspot/igv/hierarchicallayout/LayoutGraph.java:2]:
```suggestion
 * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
```

**Review by chhagedorn — APPROVED** [review]
Great work! This is a really nice feature. I've tested it on Linux and it works as expected.

**Pull request closed without merging**